intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+applications with the `torch.distributed.rpc` package. Source code of the two
+examples can be found in `PyTorch examples <https://github.com/pytorch/examples>`__
+
+Previous tutorials described `DistributedDataParallel <https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html>`__
+which supports a specific training paradigm where the model is replicated across
+multiple processes and each process handles a split of the input data.
+Sometimes, you might run into scenarios that require different training
+paradigms:
+
+1) In reinforcement learning, it might be relatively expensive to acquire
+   training data from environments while the model itself can be quite small. In
+   this case, it might be useful to spawn multiple observers running in parallel
+   and share a single agent. In this case, the agent takes care of the training
+   locally, but the application would still need libraries to send and receive
+   data between observers and the trainer
+2) Your model might be too large to fit in GPUs on a single machine, and hence
+   would need a library to help split a model onto multiple machines. Or you
+   might be implementing a parameter server training framework, where model
+   parameters and trainers live on different machines.
+
+
+The `torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ package
+can help with the above scenarios. In case 1, `RPC <https://pytorch.org/docs/master/rpc.html#rpc>`__
+and `RRef <https://pytorch.org/docs/master/rpc.html#rref>`__ can help send data
+from one worker to another and also easily referencing remote data objects. In
+case 2, `distributed autograd <https://pytorch.org/docs/master/rpc.html#distributed-autograd-framework>`__
+and `distributed optimizer <https://pytorch.org/docs/master/rpc.html#module-torch.distributed.optim>`__
+allows executing backward and optimizer step as if it is local training. In the
+next two sections, we will demonstrate APIs of
+`torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ using a
+reinforcement learning example and a language model example. Please note, this
+tutorial is not aiming at building the most accurate or efficient models to
+solve given problems, instead the main goal is to show how to use the
+`torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ package to
+build distributed training applications.
+
+
+
+Distributed Reinforcement Learning using RPC and RRef
+-----------------------------------------------------
+
+This section describes steps to build a toy distributed reinforcement learning
+model using RPC to solve CartPole-v1 from `OpenAI Gym <https://gym.openai.com>`__.
+The policy code is mostly borrowed from the existing single-thread
+`example <https://github.com/pytorch/examples/blob/master/reinforcement_learning>`__
+as shown below. We will skip details of the ``Policy`` design, and focus on RPC
+usages.
+
+.. code:: python
+
+    class Policy(nn.Module):
```

Comment:
general q: Should the tutorials be "copy pastable", i.e. should have all the imports included and some instructions on what's needed to actually be able to run the code? I've found it useful in some pytorch tutorials where I can just copy + paste the code and it works, but not all tutorials are setup this way.

Reply:
Good point! I agree, let me add imports here. Ideally, I hope we could write notebooks instead of rst for tutorials, but I recall that I run into sphinx/multiprocessing/ipython error (see #478). Let me try if that bug is fixed. If yes, I will use a notebook.